avr-hal-generic/src/serial.rs

```diff
@@ -1,5 +1,107 @@
 //! Serial Implementations
 
+use core::cmp::Ordering;
+
+// Clock is needed because the calculations needs to take core clock into account
+#[derive(Debug, Clone, Copy)]
+pub struct Baudrate<CLOCK> {
+    pub ubrr: u16,
+    pub u2x: bool,
+    pub _clock: ::core::marker::PhantomData<CLOCK>,
+}
+
+impl<CLOCK: crate::clock::Clock> PartialEq for Baudrate<CLOCK> {
+    fn eq(&self, other: &Self) -> bool {
+        self.compare_value() == other.compare_value()
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Eq for Baudrate<CLOCK> {}
+
+impl<CLOCK: crate::clock::Clock> PartialOrd for Baudrate<CLOCK> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.compare_value().cmp(&other.compare_value()))
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Ord for Baudrate<CLOCK> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.compare_value().cmp(&self.compare_value())
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> From<u32> for Baudrate<CLOCK> {
+    fn from(baud: u32) -> Self {
+        Baudrate::new(baud)
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Baudrate<CLOCK> {
+    pub fn new(baud: u32) -> Baudrate<CLOCK> {
+        let mut ubrr = (CLOCK::FREQ / 4 / baud - 1) / 2;
+        let mut u2x = true;
+        debug_assert!(ubrr <= u16::MAX as u32);
+        if ubrr > 4095 {
+            u2x = false;
+            ubrr = (CLOCK::FREQ / 8 / baud - 1) / 2;
+        }
+
+        Baudrate {
+            ubrr: ubrr as u16,
+            u2x: u2x,
+            _clock: ::core::marker::PhantomData,
+        }
+    }
+
+    pub fn with_exact(u2x: bool, ubrr: u16) -> Baudrate<CLOCK> {
+        Baudrate {
+            ubrr: ubrr,
+            u2x,
+            _clock: ::core::marker::PhantomData,
+        }
```

Comment:
Can be written a bit more concise:

```suggestion
        Baudrate {
            ubrr, u2x, _clock: ::core::marker::PhantomData,
        }
```